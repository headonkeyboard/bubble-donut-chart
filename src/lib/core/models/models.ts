export interface Point {
  x: number;
  y: number;
}

export interface RawData {
  id: string;
  label: string;
  weight: number;
  group: string;
}

export interface Section {
  id: string;
  ratio: number;
  bubbles: BubbleWithCoordsAndRadius[];
  startAngle: number;
  endAngle: number;
  gridLength: number;
  edges: GridEdges;
}

export interface GridEdges {
  top: Point[];
  right: Point[];
  bottom: Point[];
  left: Point[];
}

export type Coord = Point & { r: number };
export type BubbleWithCoordsAndRadius = RawData & Coord;
export type Grid = Array<number>;
