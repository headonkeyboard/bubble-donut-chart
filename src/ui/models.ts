export interface Section {
    bubbleCount: number;
    bubbleMinValue: number;
    bubbleMaxValue: number;
}
