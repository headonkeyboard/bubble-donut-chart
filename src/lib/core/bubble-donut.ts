import {
  RawData,
  BubbleWithCoordsAndRadius,
  Grid,
  GridEdges,
  Point,
  DonutSection,
  MaxDistancePoint,
} from "./models";
import { gridPointPositionToPixels } from "./utils/coords.utils";
import { generateGrid, getBubblePositionAndRadius } from "./utils/bubble.utils";
import { getRawDataByGroup, getRawDataWeightSum } from "./utils/raw-data.utils";

class BubbleDonut {
  sections = new Map<string, DonutSection>();
  rawData: RawData[] = [];

  donutArea = -1;
  allBubbleWeightSum = 0;

  constructor(
    public innerRadius: number,
    public outerRadius: number,
    rawData: RawData[] = []
  ) {
    this.donutArea = this.getDonutArea();

    if (rawData.length > 0) {
      this.loadRawData(rawData);
    }
  }

  /**
   * Format raw data to bubble sections
   *
   * @param rawData entries that will be formatted as Bubble Donut
   */
  loadRawData(rawData: RawData[]): void {
    this.sections.clear();

    this.rawData = rawData;

    // bubble positioning algorithm needs data to be sorted by their weight with biggest weight first
    this.rawData.sort((a, b) => b.weight - a.weight);

    // each bubble radius is proportional to the sum of all data weight
    this.allBubbleWeightSum = getRawDataWeightSum(this.rawData);

    const rawDataByGroup = getRawDataByGroup(this.rawData);

    // keep group order consistent to always display groups in the same place of the bubble donut
    const groupKeys = Array.from(rawDataByGroup.keys());
    groupKeys.sort();

    groupKeys.forEach((groupKey) => {
      const section = this.newDonutSection(
        groupKey,
        rawDataByGroup.get(groupKey) as RawData[]
      );
      this.sections.set(groupKey, section);
    });
  }

  /**
   * Returns a flat array of BubbleWithCoordsAndRadius extracted from each donut sections
   */
  getBubbles(): BubbleWithCoordsAndRadius[] {
    const bubbles: BubbleWithCoordsAndRadius[] = [];

    for (let [, bubbleSection] of this.sections) {
      bubbleSection.bubbles.forEach((bubble) => {
        bubbles.push(bubble);
      });
    }

    return bubbles;
  }

  /**
   * Return a newly created Donut Section
   *
   * @param groupKey group value from rawData entries
   * @param sectionRawData entries that will be formatted as Bubble Donut
   */
  private newDonutSection(
    groupKey: string,
    sectionRawData: RawData[]
  ): DonutSection {
    const ratio = getRawDataWeightSum(sectionRawData) / this.allBubbleWeightSum;
    const previousSections = Array.from(this.sections);

    let startAngle = 0;

    if (previousSections.length > 0) {
      startAngle = previousSections[previousSections.length - 1][1].endAngle;
    }

    const endAngle = startAngle + ratio * 2 * Math.PI;
    const gridLength = Math.min(Math.max(sectionRawData.length, 10), 100);
    const gridEdges = this.getGridEdges(gridLength, startAngle, endAngle);
    const bubblesWithCoordsAndRadius: BubbleWithCoordsAndRadius[] = [];

    let grid: Grid, maxDistancePoint: MaxDistancePoint;

    const gridPointPositionsCache = new Map<number, Point>();

    // each grid point will be mapped to it's min distance with an edge of the section
    ({ grid, maxDistancePoint } = generateGrid(
      gridEdges,
      gridLength,
      startAngle,
      endAngle,
      this.innerRadius,
      this.outerRadius,
      gridPointPositionsCache
    ));

    sectionRawData.forEach((bubble) => {
      // find a point on the grid which have the max available distance from any obstacle
      // this will be the center of the bubble
      const maxDistancePosition = gridPointPositionToPixels(
        maxDistancePoint.x,
        maxDistancePoint.y,
        startAngle,
        endAngle,
        gridLength,
        this.innerRadius,
        this.outerRadius
      );

      // apply a bit of randomness from grid position
      const { bubbleX, bubbleY, bubbleR } = getBubblePositionAndRadius(
        maxDistancePoint.maxDistance,
        maxDistancePosition,
        bubble.weight,
        this.donutArea,
        this.allBubbleWeightSum
      );

      // store newly "added" bubble to compute distance from it on next grid refresh
      bubblesWithCoordsAndRadius.push({
        ...bubble,
        x: bubbleX,
        y: bubbleY,
        r: bubbleR,
      });

      // refresh grid to map items with their min distance from any obstacle (an edge or another bubble)
      ({ grid, maxDistancePoint } = generateGrid(
        gridEdges,
        gridLength,
        startAngle,
        endAngle,
        this.innerRadius,
        this.outerRadius,
        gridPointPositionsCache,
        grid,
        bubblesWithCoordsAndRadius
      ));
    });

    return {
      id: groupKey,
      ratio,
      bubbles: bubblesWithCoordsAndRadius,
      startAngle,
      endAngle,
      gridLength,
      edges: gridEdges,
    };
  }

  /**
   * Returns many points all around a section.
   *
   * @param gridLength
   * @param startAngle
   * @param endAngle
   */
  private getGridEdges(
    gridLength: number,
    startAngle: number,
    endAngle: number
  ): GridEdges {
    const gridEdges: GridEdges = {
      top: [],
      right: [],
      bottom: [],
      left: [],
    };

    for (let i = 0; i <= gridLength; i++) {
      gridEdges.left.push(
        gridPointPositionToPixels(
          i,
          0,
          startAngle,
          endAngle,
          gridLength,
          this.innerRadius,
          this.outerRadius
        )
      ); // side edge

      gridEdges.right.push(
        gridPointPositionToPixels(
          i,
          gridLength,
          startAngle,
          endAngle,
          gridLength,
          this.innerRadius,
          this.outerRadius
        )
      ); // side edge

      // reduce some edges to reduce compute time during grid generation
      gridEdges.bottom.push(
        gridPointPositionToPixels(
          0,
          i,
          startAngle,
          endAngle,
          gridLength,
          this.innerRadius,
          this.outerRadius
        )
      ); // inner edge

      gridEdges.top.push(
        gridPointPositionToPixels(
          gridLength,
          i,
          startAngle,
          endAngle,
          gridLength,
          this.innerRadius,
          this.outerRadius
        )
      ); // outer edge
    }

    return gridEdges;
  }

  /**
   * Returns donut area by subtracting the area of the outer disc with the area of the inner disc
   */
  private getDonutArea(): number {
    const externalCircleArea = Math.PI * Math.pow(this.outerRadius, 2);
    const innerCircleArea = Math.PI * Math.pow(this.innerRadius, 2);

    return externalCircleArea - innerCircleArea;
  }
}

export { BubbleDonut };
