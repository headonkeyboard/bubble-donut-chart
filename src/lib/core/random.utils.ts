import { RawData } from "./models/models";

export function getBubbles(bubbleCount: number, sections: number[]): RawData[] {
  const bubbles: RawData[] = [];

  let nbBubble = 0;

  sections.forEach((section, sectionIndex) => {
    const sectionBubbleCount = Math.floor((section / 100) * bubbleCount);
    for (let j = 0; j < sectionBubbleCount; j++) {
      let bubbleValue = getRandomInt(1, 10);

      const bubbleId = `bubble_${nbBubble}`;

      bubbles.push({
        group: `group ${sectionIndex + 1}`,
        weight: bubbleValue,
        id: bubbleId,
        label: "some label",
      });

      nbBubble++;
    }
  });

  return bubbles;
}

export function getRandomInt(min: number, max: number): number {
  return Math.floor(Math.random() * Math.floor(max - min)) + min;
}
