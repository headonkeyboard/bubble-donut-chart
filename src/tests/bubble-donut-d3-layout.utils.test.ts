import React from "react";
import {
  createD3SvgRoot,
  layoutBubbles, rawDataToBubbleDonut,
  resizeD3ViewportAndBubbles,
} from "../ui/bubble-donut-chart/utils/bubble-donut-d3-layout.utils";
import {BubbleWithCoordsAndRadius} from "../lib/core/models";
import dataSample from "./fixtures/data-sample.json";

const SVG_ROOT_SIZE = 200;

let wrapper!: HTMLDivElement;
let selection!: d3.Selection<
  SVGSVGElement,
  BubbleWithCoordsAndRadius,
  null,
  undefined
>;

describe("bubble donut d3 layout utils", () => {
  const bubbleDonut = rawDataToBubbleDonut(dataSample);
  const bubbles = bubbleDonut.getBubbles();
  const groupCount = bubbleDonut.sections.size;

  beforeEach(() => {
    wrapper = document.createElement("div");
    selection = createD3SvgRoot(wrapper, SVG_ROOT_SIZE);
  });

  test("createD3SvgRoot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("layoutBubbles", () => {
    layoutBubbles(selection, SVG_ROOT_SIZE, bubbles, groupCount);

    const selectionNodeChildren = selection.node()?.children;
    expect(selectionNodeChildren?.length).toBe(dataSample.length);

    for (let i = 0; i < selectionNodeChildren.length; i++) {
      const child = selectionNodeChildren.item(i);
      expect(child.tagName).toBe("circle");
      expect(child.getAttribute("fill")).toBeDefined();
      expect(child.getAttribute("opacity")).toBeDefined();
      expect(child.getAttribute("r")).toBeDefined();
      expect(child.getAttribute("cx")).toBeDefined();
      expect(child.getAttribute("cy")).toBeDefined();
    }
  });

  test("resizeD3ViewportAndBubbles should resize svg root element and circles children", (done) => {
    const RESIZE_RATIO = 2;

    layoutBubbles(selection, SVG_ROOT_SIZE, bubbles, groupCount);

    type Circle = {
      r: number;
      cx: number;
      cy: number;
    };

    const circlePropertiesBefreResize = new Map<number, Circle>();

    // wait for layout animations to finish
    setTimeout(() => {
      const selectionNodeChildren = selection?.node()
        .children as HTMLCollection;

      for (let i = 0; i < selectionNodeChildren.length; i++) {
        const child = selectionNodeChildren.item(i) as Element;
        circlePropertiesBefreResize.set(i, {
          cx: parseFloat(child.getAttribute("cx") as string),
          cy: parseFloat(child.getAttribute("cy") as string),
          r: parseFloat(child.getAttribute("r") as string),
        });
      }

      resizeD3ViewportAndBubbles(selection, SVG_ROOT_SIZE * RESIZE_RATIO);

      for (let i = 0; i < selectionNodeChildren.length; i++) {
        const circleBeforeResize = circlePropertiesBefreResize.get(i) as Circle;

        const child = selectionNodeChildren.item(i) as Element;
        const childR = parseFloat(child.getAttribute("r") as string);
        const childCx = parseFloat(child.getAttribute("cx") as string);
        const childCy = parseFloat(child.getAttribute("cy") as string);

        expect(childR).toBe(circleBeforeResize.r * RESIZE_RATIO);
        expect(childCx).toBe(circleBeforeResize.cx * RESIZE_RATIO);
        expect(childCy).toBe(circleBeforeResize.cy * RESIZE_RATIO);
      }

      done();
    }, 1000);
  });
});
