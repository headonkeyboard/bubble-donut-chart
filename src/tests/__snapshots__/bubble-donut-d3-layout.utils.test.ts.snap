// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bubble donut d3 layout utils createD3SvgRoot 1`] = `
<div>
  <svg
    height="200"
    width="200"
  />
</div>
`;
